client/network/sync/src/lib.rs

```diff
@@ -578,12 +578,19 @@ where
 
 	/// Returns the current sync status.
 	pub fn status(&self) -> Status<B> {
-		let best_seen = self.peers.values().map(|p| p.best_number).max();
+		let best_seen = if self.peers.is_empty() {
+			None
+		} else {
+			Some(
+				self.peers.values().fold(NumberFor::<B>::from(0u32), |c, p| c + p.best_number) /
+					self.peers.len().saturated_into::<u32>().into(),
```

Comment:
Ideally this should be median or mode rather than average.

Reply:
mean is trivially attackable and not any better than we already have. median works, though we'd still be susceptible to a 51% attack of peers. probably a bit more resilient would be weight the median in favour of any known validator peers (i.e. treat every peer who is a known validator as 5 reports rather than just 1).

given we know block time and we have a most-recent block & timestamp stored already and we have a local clock, we can calculate the approximate *maximum* chain height (assuming that a block is produced every 6 seconds). we can disregard reported heights which exceed this.

Reply:
Peers that can't provide advertised best block will be banned eventually.

We should also configure validator by default to use higher number of outgoing connections and lower number of accepted incoming connections, which should help with eclipse attacks in general.

> (i.e. treat every peer who is a known validator as 5 reports rather than just 1).

We don't really know which peers are validators on the sync protocol level. Validators would need  to sign the handshake with their key.

> given we know block time and we have a most-recent block & timestamp stored already and we have a local clock, we can calculate the approximate maximum chain height (assuming that a block is produced every 6 seconds). we can disregard reported heights which exceed this.

This could work, although approximation would need to be done from the last block, rather than genesis. Consider a peer reporting best block 10,000,100  while true best block is 10,000,000. It is within plausible range if approximating from genesis, but we still want to discard it eventually. An accurate estimation on wheather `best_block` is true is possible once we are at the tip of the chain. Also this would also require the sync protocol to know about block timestamps.